src/sqlfluff/rules/L016.py

```diff
@@ -28,620 +23,144 @@ class Rule_L016(BaseRule):
+                    ):
+                        self.logger.debug(
+                            "Purging result with template comment line: %s",
+                            res.anchor.pos_marker.working_line_no,
                         )
-                    )
-                    return fixes
-                raise ValueError(
-                    f"Unexpected break generated at {self}"
-                )  # pragma: no cover
+                        results.remove(res)
+                        break
 
-        segment_buff: Tuple[RawSegment, ...] = ()
-        whitespace_buff: Tuple[RawSegment, ...] = ()
-        indent_impulse = 0
-        indent_balance = 0
-        is_pause = False
-
-        seg: RawSegment
-        for seg in segments:
-            if indent_section is None:
-                if seg.is_type("whitespace") or seg.is_meta:
-                    whitespace_buff += (seg,)
-                else:
-                    indent_section = Section(
-                        segments=whitespace_buff,
-                        role="indent",
-                        indent_balance=indent_balance,
-                    )
-                    whitespace_buff = ()
-                    segment_buff = (seg,)
-            else:
-                if seg.is_type("whitespace") or seg.is_meta:
-                    whitespace_buff += (seg,)
-                    if seg.is_meta:
-                        indent_impulse += seg.indent_val
-                else:
-                    # We got something other than whitespace or a meta.
-                    # Have we passed an indent?
-                    if indent_impulse != 0:
-                        # Yes. Bank the section, perhaps also with a content
-                        # section.
-                        if segment_buff:
-                            chunk_buff.append(
-                                Section(
-                                    segments=segment_buff,
-                                    role="content",
-                                    indent_balance=indent_balance,
+        # Ignore any comment clauses if present.
+        if self.ignore_comment_clauses:
+            raw_segments = context.segment.raw_segments
+            for res in results[:]:
+                # The anchor should be the first raw on the line. Work forward
+                # until we're not on the line. Check if any have a parent which
+                # is a comment_clause.
+                raw_idx = raw_segments.index(res.anchor)
+                for _, seg in enumerate(raw_segments[raw_idx:], raw_idx):
+                    path = context.segment.path_to(seg)
+                    if (
+                        seg.pos_marker.working_line_no
+                        != res.anchor.pos_marker.working_line_no
+                    ):
+                        # We've gone past the end of the line. Stop looking.
+                        break
```

Comment:
Hmmmm - good idea. I'll investigate that option.

Reply:
Because of the way the code is structed, I actually think I'll leave it like this. I tried a couple of ways of re-doing it, but I think while they avoid the duplication they're harder to understand and reason about. This way is at least clear even if it does involve a little duplication.

Reply:
No problem, it was kind of a nitpick.